exploration/0474-text-vs-code.md

```diff
@@ -0,0 +1,402 @@
+
+## Background
+
+Existing message and template formatting languages tend to start in "text" mode,
+and require special syntax like `{{` or `{%` to enter "code" mode.
+
+ICU MessageFormat and Fluent both support inline selectors
+separated from the text using `{ ... }` for multi-variant messages.
+
+[Mustache templates](https://mustache.github.io/mustache.5.html)
+and related languages wrap "code" in `{{ ... }}`.
+In addition to placeholders that are replaced by their interpolated value during formatting,
+this also includes conditional blocks using `{{#...}}`/`{{/...}}` wrappers.
+
+[Handlebars](https://handlebarsjs.com/guide/) extends Mustache expressions
+with operators such as `{{#if ...}}` and `{{#each ...}}`,
+as well as custom formatting functions that become available as e.g. `{{bold ...}}`.
+
+[Jinja templates](https://jinja.palletsprojects.com/en/3.1.x/templates/) separate
+`{% statements %}` and `{{ expressions }}` from the base text.
+The former may define tests that determine the inclusion of subsequent text blocks in the output.
+
+A cost that the message formatting and templating languages mentioned above need to rely on
+is some rule or behaviour that governs how to deal with whitespace at the beginning and end of a pattern,
+as statements may be separated from each other by newlines or other constructs for legibility.
+
+Other formats supporting multiple message variants tend to rely on a surrounding resource format to define variants,
+such as [Rails internationalization](https://guides.rubyonrails.org/i18n.html#pluralization) in Ruby or YAML
+and [Android String Resources](https://developer.android.com/guide/topics/resources/string-resource.html#Plurals) in XML.
+These formats rely on the resource format providing clear delineation of the beginning and end of a pattern.
+
+## Use-Cases
+
+Most messages in any localization system do not contain any expressions, statements or variants.
+These should be expressible as easily as possible.
+
+Many messages include expressions that are to be interpolated during formatting.
+For example, a greeting like "Hello, user!" may be formatted in many locales with the `user`
+being directly set by an input variable.
+
+Sometimes, interpolated values need explicit formatting within a message.
+For example, formatting a message like "You have eaten 3.2 apples"
+may require the input numerical value
+to be formatted with an explicit `minimumFractionDigits` option.
+
+Some messages require multiple variants.
+This is often related to plural cases, such as "You have 3 new messages",
+where the value `3` is an input and the "messages" needs to correspond with its plural category.
+
+Rarely, messages needs to include leading or trailing whitespace due to
+e.g. how they will be concatenated with other text,
+or as a result of being segmented from some larger volume of text.
+
+## Requirements
+
+Easy things should be easy, and hard things should be possible.
```

Comment:
Majority opinion seems to be that standard phrasing of the Alan Kay quote is "Simple things should be simple, complex things should be possible." You might also consider the alternate form proposed by our illustrious reviewer: https://twitter.com/LeaVerou/status/1701545112583500266
> I often paraphrase it: “Common things should be easy, uncommon things should be possible”.

Reply:
```suggestion
Common things should be easy, uncommon things should be possible.
```